File: 4-kyu/9-Next-bigger-number-with-the-same-digits/go/main.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func main() {
	fmt.Println(NextBiggerA(144))
	fmt.Println(NextBigger(414))
	fmt.Println(NextBigger(12))
	fmt.Println(NextBigger(2017))
	fmt.Println(NextBigger(531))
	fmt.Println(NextBigger(111))
	fmt.Println(NextBigger(11112))
	fmt.Println(NextBigger(59884848459853))
}

// NextBigger :
func NextBigger(n int) int {
	if n < 12 {
		return -1
	}
	strbase := strconv.Itoa(n)
	// can't be rearranged
	canBe := false
	for i := 0; i < len(strbase)-1; i++ {
		if strbase[i] < strbase[i+1] {
			canBe = true
			break
		}
	}
	if !canBe {
		return -1
	}
	// can't be rearranged
	for nPlus := n + 1; ; nPlus++ {
		strPlus := strconv.Itoa(nPlus)
		valid := true
		for _, v := range strbase {
			if strings.Count(strbase, string(v)) != strings.Count(strPlus, string(v)) {
				valid = false
				break
			}
		}
		if !valid {
			continue
		}
		return nPlus
	}
}

///================ other practices ==================///

func NextBiggerA(n int) (r int) {
	s := []rune(strconv.Itoa(n))

	i := len(s) - 1
	for i > 0 && s[i-1] >= s[i] {
		i--
	}
	if i == 0 {
		return -1
	}

	j := len(s) - 1
	for s[j] <= s[i-1] {
		j--
	}
	s[i-1], s[j] = s[j], s[i-1]

	for k, m := i, len(s)-1; k < m; k, m = k+1, m-1 {
		s[k], s[m] = s[m], s[k]
	}
	r, _ = strconv.Atoi(string(s))
	return
}

File: 4-kyu/9-Next-bigger-number-with-the-same-digits/python/main.py

def next_bigger(n):
    if n < 12 :
        return -1
    can_be = False
    i = 0
    str_base = str(n)
    while i < len(str_base)-1:
        if str_base[i] < str_base[i+1]:
            can_be = True
            break
        i += 1
    if not can_be:
        return -1
    n_plus = n
    while True:
        n_plus +=1
        str_plus = str(n_plus)
        valid = True
        for v in str_base:
            if str_base.count(v) != str_plus.count(v):
                valid = False
                break
        if not valid:
            continue
        return n_plus

### ================ other practices ================== 


import itertools
def next_biggerA(n):
    s = list(str(n))
    for i in range(len(s)-2,-1,-1):
        if s[i] < s[i+1]:
            t = s[i:]
            m = min(filter(lambda x: x>t[0], t))
            t.remove(m)
            t.sort()
            s[i:] = [m] + t
            return int("".join(s))
    return -1
